### PR TITLE
Ref count search response bytes

### LINKED ISSUE #102657: [CI] BasicDistributedJobsIT testMaxConcurrentJobAllocations failing
The failure indicates a leaked multi-search response. I took a quick look and see several usages of `multiSearch` in the ML codebase which do not appear to close their responses. Not sure which one caused this but I guess they all need fixing. Relates #102604.

**Build scan:**
https://gradle-enterprise.elastic.co/s/3viumbl72hyti/tests/:x-pack:plugin:ml:internalClusterTest/org.elasticsearch.xpack.ml.integration.BasicDistributedJobsIT/testMaxConcurrentJobAllocations

**Reproduction line:**
```
./gradlew ':x-pack:plugin:ml:internalClusterTest' --tests "org.elasticsearch.xpack.ml.integration.BasicDistributedJobsIT.testMaxConcurrentJobAllocations" -Dtests.seed=88982C83B70D110B -Dtests.locale=it -Dtests.timezone=Navajo -Druntime.java=21
```

**Applicable branches:**
main

**Reproduces locally?:**
Didn't try

**Failure history:**
https://es-delivery-stats.elastic.dev/app/dashboards#/view/dcec9e60-72ac-11ee-8f39-55975ded9e63?_g=(refreshInterval:(pause:!t,value:60000),time:(from:now-7d%2Fd,to:now))&_a=(controlGroupInput:(chainingSystem:HIERARCHICAL,controlStyle:twoLine,ignoreParentSettings:(ignoreFilters:!f,ignoreQuery:!f,ignoreTimerange:!f,ignoreValidations:!t),panels:('0c0c9cb8-ccd2-45c6-9b13-96bac4abc542':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:task.keyword,grow:!t,id:'0c0c9cb8-ccd2-45c6-9b13-96bac4abc542',searchTechnique:wildcard,selectedOptions:!(),singleSelect:!t,title:'Gradle%20Task',width:medium),grow:!t,order:0,type:optionsListControl,width:small),'144933da-5c1b-4257-a969-7f43455a7901':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:name.keyword,grow:!t,id:'144933da-5c1b-4257-a969-7f43455a7901',searchTechnique:wildcard,selectedOptions:!('testMaxConcurrentJobAllocations'),title:Test,width:medium),grow:!t,order:2,type:optionsListControl,width:medium),'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:className.keyword,grow:!t,id:'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850',searchTechnique:wildcard,selectedOptions:!('org.elasticsearch.xpack.ml.integration.BasicDistributedJobsIT'),title:Suite,width:medium),grow:!t,order:1,type:optionsListControl,width:medium))))

**Failure excerpt:**
```
java.lang.AssertionError: 
Expected: an empty collection
     but: <[LEAK: resource was not cleaned up before it was garbage-collected.
Recent access records: 
Created at:
	org.elasticsearch.action.search.ArraySearchPhaseResults.<init>(ArraySearchPhaseResults.java:25)
	org.elasticsearch.action.search.QueryPhaseResultConsumer.<init>(QueryPhaseResultConsumer.java:85)
	org.elasticsearch.action.search.SearchPhaseController.newSearchPhaseResults(SearchPhaseController.java:788)
	org.elasticsearch.action.search.TransportSearchAction$AsyncSearchActionProvider.newSearchPhase(TransportSearchAction.java:1286)
	org.elasticsearch.action.search.TransportSearchAction.executeSearch(TransportSearchAction.java:1115)
	org.elasticsearch.action.search.TransportSearchAction.executeLocalSearch(TransportSearchAction.java:897)
	org.elasticsearch.action.search.TransportSearchAction.lambda$executeRequest$10(TransportSearchAction.java:333)
	org.elasticsearch.action.ActionListenerImplementations$ResponseWrappingActionListener.onResponse(ActionListenerImplementations.java:236)
	org.elasticsearch.index.query.Rewriteable.rewriteAndFetch(Rewriteable.java:109)
	org.elasticsearch.index.query.Rewriteable.rewriteAndFetch(Rewriteable.java:77)
	org.elasticsearch.action.search.TransportSearchAction.executeRequest(TransportSearchAction.java:445)
	org.elasticsearch.action.search.TransportSearchAction.doExecute(TransportSearchAction.java:300)
	org.elasticsearch.action.search.TransportSearchAction.doExecute(TransportSearchAction.java:112)
	org.elasticsearch.action.support.TransportAction$RequestFilterChain.proceed(TransportAction.java:87)
	org.elasticsearch.action.support.ActionFilter$Simple.apply(ActionFilter.java:53)
	org.elasticsearch.action.support.TransportAction$RequestFilterChain.proceed(TransportAction.java:85)
	org.elasticsearch.action.support.TransportAction.execute(TransportAction.java:62)
	org.elasticsearch.tasks.TaskManager.registerAndExecute(TaskManager.java:197)
	org.elasticsearch.client.internal.node.NodeClient.executeLocally(NodeClient.java:108)
	org.elasticsearch.client.internal.node.NodeClient.doExecute(NodeClient.java:86)
	org.elasticsearch.client.internal.support.AbstractClient.execute(AbstractClient.java:382)
	org.elasticsearch.client.internal.support.AbstractClient.search(AbstractClient.java:517)
	org.elasticsearch.action.search.TransportMultiSearchAction.executeSearch(TransportMultiSearchAction.java:154)
	org.elasticsearch.action.search.TransportMultiSearchAction$1.handleResponse(TransportMultiSearchAction.java:180)
	org.elasticsearch.action.search.TransportMultiSearchAction$1.onFailure(TransportMultiSearchAction.java:165)
	org.elasticsearch.action.ActionListenerImplementations.safeAcceptException(ActionListenerImplementations.java:62)
	org.elasticsearch.action.ActionListenerImplementations.safeOnFailure(ActionListenerImplementations.java:73)
	org.elasticsearch.action.ActionListener$4.onFailure(ActionListener.java:341)
	org.elasticsearch.tasks.TaskManager$1.onFailure(TaskManager.java:215)
	org.elasticsearch.action.ActionListenerImplementations.safeAcceptException(ActionListenerImplementations.java:62)
	org.elasticsearch.action.ActionListenerImplementations.safeOnFailure(ActionListenerImplementations.java:73)
	org.elasticsearch.action.DelegatingActionListener.onFailure(DelegatingActionListener.java:27)
	org.elasticsearch.action.ActionListenerImplementations.safeAcceptException(ActionListenerImplementations.java:62)
	org.elasticsearch.action.ActionListenerImplementations.safeOnFailure(ActionListenerImplementations.java:73)
	org.elasticsearch.action.DelegatingActionListener.onFailure(DelegatingActionListener.java:27)
	org.elasticsearch.action.ActionListenerImplementations$RunAfterActionListener.onFailure(ActionListenerImplementations.java:269)
	org.elasticsearch.action.ActionListenerImplementations.safeAcceptException(ActionListenerImplementations.java:62)
	org.elasticsearch.action.ActionListenerImplementations.safeOnFailure(ActionListenerImplementations.java:73)
	org.elasticsearch.action.ActionListener$4.onFailure(ActionListener.java:341)
	org.elasticsearch.action.search.AbstractSearchAsyncAction.raisePhaseFailure(AbstractSearchAsyncAction.java:732)
	org.elasticsearch.action.search.AbstractSearchAsyncAction.onPhaseFailure(AbstractSearchAsyncAction.java:709)
	org.elasticsearch.action.search.AbstractSearchAsyncAction.executeNextPhase(AbstractSearchAsyncAction.java:401)
	org.elasticsearch.action.search.AbstractSearchAsyncAction.onPhaseDone(AbstractSearchAsyncAction.java:741)
	org.elasticsearch.action.search.AbstractSearchAsyncAction.onShardFailure(AbstractSearchAsyncAction.java:494)
	org.elasticsearch.action.search.AbstractSearchAsyncAction$1.onFailure(AbstractSearchAsyncAction.java:332)
	org.elasticsearch.action.ActionListenerImplementations.safeAcceptException(ActionListenerImplementations.java:62)
	org.elasticsearch.action.ActionListenerImplementations.safeOnFailure(ActionListenerImplementations.java:73)
	org.elasticsearch.action.DelegatingActionListener.onFailure(DelegatingActionListener.java:27)
	org.elasticsearch.action.ActionListenerResponseHandler.handleException(ActionListenerResponseHandler.java:54)
	org.elasticsearch.action.search.SearchTransportService$ConnectionCountingHandler.handleException(SearchTransportService.java:553)
	org.elasticsearch.transport.TransportService$UnregisterChildTransportResponseHandler.handleException(TransportService.java:1703)
	org.elasticsearch.transport.TransportService$ContextRestoreResponseHandler.handleException(TransportService.java:1420)
	org.elasticsearch.transport.TransportService$7.doRun(TransportService.java:1304)
	org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:983)
	org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)
	java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	java.base/java.lang.Thread.run(Thread.java:1583)]>

  at __randomizedtesting.SeedInfo.seed([88982C83B70D110B:FD6BD27394E3FDD8]:0)
  at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
  at org.junit.Assert.assertThat(Assert.java:956)
  at org.junit.Assert.assertThat(Assert.java:923)
  at org.elasticsearch.test.ESTestCase.checkStaticState(ESTestCase.java:721)
  at org.elasticsearch.test.ESTestCase.after(ESTestCase.java:512)
  at jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
  at java.lang.reflect.Method.invoke(Method.java:580)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1758)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:1004)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleSetupTeardownChained$1.evaluate(TestRuleSetupTeardownChained.java:48)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at org.apache.lucene.tests.util.TestRuleThreadAndTestName$1.evaluate(TestRuleThreadAndTestName.java:45)
  at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
  at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.forkTimeoutingTask(ThreadLeakControl.java:843)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$3.evaluate(ThreadLeakControl.java:490)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.runSingleTest(RandomizedRunner.java:955)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$5.evaluate(RandomizedRunner.java:840)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$6.evaluate(RandomizedRunner.java:891)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$7.evaluate(RandomizedRunner.java:902)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleStoreClassName$1.evaluate(TestRuleStoreClassName.java:38)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleAssertionsRequired$1.evaluate(TestRuleAssertionsRequired.java:53)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
  at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
  at org.apache.lucene.tests.util.TestRuleIgnoreTestSuites$1.evaluate(TestRuleIgnoreTestSuites.java:47)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.lambda$forkTimeoutingTask$0(ThreadLeakControl.java:850)
  at java.lang.Thread.run(Thread.java:1583)

```

### PR DESCRIPTION
**Pull request by original-brownbear**
Final step in  #102030 ... actually makes `SearchHit` read a releasable bytes reference.
Does still fallback to copying to unrolled buffers here and there which can be removed in follow-ups where it's worth the effort (aggs being the most important one probably).

Hard to create very reliable benchmarks for this because all our macro-benchmarks are quite noisy. Running http logs and PMC though, there's a statistically significant reduction in GC and reduced tail latencies in most benchmarks.

The overhead for ref-counting these bytes isn't visible in profiling as far as I can tell and for large source values, no corresponding large `byte[]` are created any longer outside of the few remaining spots where we copy to pooled buffers.

closes #102657

### CODE REVIEW

**Comment by original-brownbear** [comment]
This is good for review now I believe. I will be adding more detail and some benchmark numbers over the next couple h though.

File: server/src/main/java/org/elasticsearch/action/search/ArraySearchPhaseResults.java [hunk 1]
```diff
@@ -22,7 +22,13 @@
 class ArraySearchPhaseResults<Result extends SearchPhaseResult> extends SearchPhaseResults<Result> {
     final AtomicArray<Result> results;
 
-    private final RefCounted refCounted = LeakTracker.wrap(AbstractRefCounted.of(this::doClose));
+    private final AtomicBoolean closed = new AtomicBoolean(false);
+
+    private final Releasable releasable = LeakTracker.wrap(() -> {
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/action/search/ArraySearchPhaseResults.java:27]: Changed plans on this one. Ref-counting was not appropriate here. We want to have this tracked by the context to be release as a last-resort but we also clearly know when this can be released earlier. So I just made it closable with the double-close noop so we can close it wherever.

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/persistence/JobResultsProvider.java [hunk 2]
```diff
@@ -217,7 +217,7 @@ public void checkForLeftOverDocuments(Job job, ActionListener<Boolean> listener)
         ActionListener<MultiSearchResponse> searchResponseActionListener = new DelegatingActionListener<>(listener) {
             @Override
             public void onResponse(MultiSearchResponse response) {
-                List<SearchHit> searchHits = new ArrayList<>();
+                List<String> searchHits = new ArrayList<>();
```
Comment (original-brownbear) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/persistence/JobResultsProvider.java:220]: This could have been pulled into its own PR, left it here for now to keep things from being blocked again. Just a trivial logical change here to reduce the logical scope of the search hits here.

**Review by benwtrent — COMMENTED** [review]
Whew! what a read. This is awesome. 

I eagerly await the performance numbers to see if there is a significant improvement. 

It is frustrating to see how  many non-test things require unPooled :(. But hopefully overtime those things can be updated to ensure we don't copy around bytes needlessly.

**Comment by benwtrent** [comment]
@original-brownbear could you post some numbers for a before/after of the benchmarks for this change?

File: modules/rank-eval/src/main/java/org/elasticsearch/index/rankeval/RatedSearchHit.java [hunk 3]
```diff
@@ -33,12 +33,12 @@ public class RatedSearchHit implements Writeable, ToXContentObject {
     private final OptionalInt rating;
 
     public RatedSearchHit(SearchHit searchHit, OptionalInt rating) {
-        this.searchHit = searchHit;
+        this.searchHit = searchHit.asUnpooled();
```
[the remark this answers is not in the document]
Comment (original-brownbear) [reply modules/rank-eval/src/main/java/org/elasticsearch/index/rankeval/RatedSearchHit.java:36]: Yes, this (like pretty much all other spots that use this method) is just a short-cut. Probably easy to fix this in a follow-up :)

File: server/src/internalClusterTest/java/org/elasticsearch/search/functionscore/RandomScoreFunctionIT.java [hunk 4]
```diff
@@ -112,7 +112,7 @@ public void testConsistentHitsWithSameSeed() throws Exception {
                             CoreMatchers.equalTo(0)
                         );
                         final int hitCount = response.getHits().getHits().length;
-                        final SearchHit[] currentHits = response.getHits().getHits();
+                        final SearchHit[] currentHits = response.getHits().asUnpooled().getHits();
```
[the remark this answers is not in the document]
Comment (original-brownbear) [reply server/src/internalClusterTest/java/org/elasticsearch/search/functionscore/RandomScoreFunctionIT.java:115]: True but it's more code/change to make that conditional with how things work today. I'd deal with this in a follow-up where I was planning to restrict the explicit use of the array backing a `SearchHits` as much as possible.

File: server/src/main/java/org/elasticsearch/action/search/FetchSearchPhase.java [hunk 5]
```diff
@@ -38,11 +38,16 @@ final class FetchSearchPhase extends SearchPhase {
...
+        this(
+            resultConsumer,
+            aggregatedDfs,
+            context,
+            (response, queryPhaseResults) -> new ExpandSearchPhase(
+                context,
+                response.hits,
+                () -> new FetchLookupFieldsPhase(context, response, queryPhaseResults)
+            )
+        );
```
[the remark this answers is not in the document]
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/action/search/FetchSearchPhase.java:50]: Yes exactly :)

File: server/src/main/java/org/elasticsearch/search/SearchHits.java [hunk 6]
```diff
@@ -63,39 +68,100 @@ public SearchHits(
...
+            hits.length == 0 ? ALWAYS_REFERENCED : LeakTracker.wrap(new AbstractRefCounted() {
+                @Override
+                protected void closeInternal() {
+                    for (int i = 0; i < hits.length; i++) {
+                        assert hits[i] != null;
+                        hits[i].decRef();
+                        hits[i] = null;
+                    }
+                }
+            })
```
[the remark this answers is not in the document]
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/search/SearchHits.java:88]: The problem is that we add loads more ref-counting and code if we incref every hit we pass to the constructor of `SearchHits` and dec-ref it where it's constructed. So I went with the pattern of taking ownership here, just like I did for the m-search response for example. I think going forward it would actually make sense to try and move more code to this pattern as an optimisation. Maybe we could even make more stuff closeable instead of ref-counted that way and find more optimisations that way :)

File: server/src/main/java/org/elasticsearch/search/SearchHit.java [hunk 7]
```diff
@@ -448,6 +497,7 @@ public String getSourceAsString() {
      * The source of the document as a map (can be {@code null}).
      */
     public Map<String, Object> getSourceAsMap() {
+        assert hasReferences();
```
[the remark this answers is not in the document]
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/search/SearchHit.java:500]: Technically yes, but I went for being strict here. We really shouldn't have code touching released instances ever. Much easier to trust tests if we verify that fact in more rather than fewer spots IMO

File: server/src/main/java/org/elasticsearch/search/SearchHit.java [hunk 8]
```diff
@@ -463,6 +513,7 @@ public Map<String, Object> getSourceAsMap() {
      * The hit field matching the given field name.
      */
     public DocumentField field(String fieldName) {
+        assert hasReferences();
```
[the remark this answers is not in the document]
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/search/SearchHit.java:516]: https://github.com/elastic/elasticsearch/pull/103763#discussion_r1453528022 same reasoning here. I'd love to do a follow-up that expands ref counting to the fields in some cases as well.

File: server/src/main/java/org/elasticsearch/search/aggregations/metrics/InternalTopHits.java [hunk 9]
```diff
@@ -52,7 +52,7 @@ public InternalTopHits(
         this.from = from;
         this.size = size;
         this.topDocs = topDocs;
-        this.searchHits = searchHits;
+        this.searchHits = searchHits.asUnpooled();
```
[the remark this answers is not in the document]
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/search/aggregations/metrics/InternalTopHits.java:55]: ++

File: server/src/main/java/org/elasticsearch/search/SearchHits.java [hunk 10]
```diff
@@ -63,39 +68,100 @@ public SearchHits(
...
         @Nullable Object[] collapseValues
+    ) {
+        this(
+            hits,
+            totalHits,
+            maxScore,
+            sortFields,
+            collapseField,
+            collapseValues,
+            hits.length == 0 ? ALWAYS_REFERENCED : LeakTracker.wrap(new AbstractRefCounted() {
```
[the remark this answers is not in the document]
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/search/SearchHits.java:79]: True, my reasoning for not adding that check would be that in production usage of this should effectively always work with ref-counted hirs and that condition would force another redundant iteration over the array on us.

File: server/src/main/java/org/elasticsearch/search/SearchHit.java [hunk 11]
```diff
@@ -653,13 +704,72 @@ public Map<String, Float> getMatchedQueriesAndScores() {
      * @return Inner hits or <code>null</code> if there are none
      */
     public Map<String, SearchHits> getInnerHits() {
+        assert hasReferences();
```
[the remark this answers is not in the document]
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/search/SearchHit.java:707]: Maybe not right? If we do that here, then we'd have to do it for the source itself before returning it and spots like that also? Seems like consistently asserting ref-counts on fields of ref-counted stuff is going to be a lot of code and doesn't add that much since whoever gets this map is probably going to do something with its contents anyway?

**Review by benwtrent — APPROVED** [review]
Lots to do yet. But this is awesome. I am mostly concerned about multi-node search testing, and CI should handle most of that. We just need to be vigilant.

**Comment by original-brownbear** [comment]
Wee thanks so much Ben! Merging with Luca's blessing! I'll be on the lookout for leaks :)!

**Review by astefan — COMMENTED** [review]
~~LGTM. Thank you, Armin.~~ Wrong PR review.
